std/stdio.d

```diff
@@ -2878,12 +2878,20 @@ $(D Range) that locks the file and allows fast writing to it.
                     return;
                 }
             }
-
-            // put each element in turn.
-            alias Elem = Unqual!(ElementType!A);
-            foreach (Elem c; writeme)
+            else static if (isSomeString!A && C.sizeof > 1)
             {
-                put(c);
+                foreach (c; writeme)
+                {
+                    put(c);
+                }
+            }
+            else
+            {
+                // put each element in turn.
+                for (; !writeme.empty; writeme.popFront)
+                {
+                    put(writeme.front);
+                }
```

Comment:
This seems to be a strange way to write the loop. What's wrong with just `foreach (e; writeme) put(e);`?

Reply:
And on that note, why not just use `foreach (c; writeme) put(e);` for both branches of the static if? In which case we could just fold it into a single generic loop.  AFAICT, the only real difference is that we have `foreach (c; writeme)` rather than `foreach (Elem c; writeme)` which would force auto-decoding in the case of a wide string. Getting rid of the original stuff surrounding `Elem` should take care of that.

(Though granted, the difference is rather subtle; a comment explaining this might be in order.)

Reply:
>This seems to be a strange way to write the loop. What's wrong with just

See the ASM here: https://github.com/dlang/phobos/pull/6091#issuecomment-361329540

An unnecessary copy of `writeme` is made regardless of optimizations.

Reply:
Ah I see.  Does gdc/ldc optimize that out?  It seems kinda sad that the construct that's supposed to natively support ranges would also introduce such blatant inefficiencies.

(Though I can understand that historically, we've always made a copy of slices when running foreach over them, so that foreach doesn't unexpectedly consume your slice. I suppose the same argument could be made for other range types as well. But the optimizer should be smart enough, at least in theory, to notice that the original value of the range argument isn't needed, so the copy could be eliminated.)

Reply:
>Does gdc/ldc optimize that out? 

I can't make heads or tails of ldc's ASM output. You can see it here https://run.dlang.io/is/ck7lsp

I'm going to assume "no", because there are cases where not making a copy could have subtle behavior changes.

>Though I can understand that historically, we've always made a copy of slices when running foreach over them, so that foreach doesn't unexpectedly consume your slice. I suppose the same argument could be made for other range types as well.

Yeah that's the idea, and it makes sense for user code to always use foreach regardless for simplicity's sake.

Reply:
Haha, it seems ldc's aggressive optimizer has defeated us once again. You might want to put the code into a public function that returns a value, because ldc's optimizer has apparently noticed that the value computed by the loop is never used, so it has completely elided the entire body of `main()` to just `return 0`. :-P

Reply:
(And I just tried changing the code to return the value computed by the loop: again, ldc's optimizer is very aggressive and decided to execute the loop at compile-time, so `main()` just returns the precomputed value directly without any computation.)

Reply:
OK, the only way I could convince ldc not to replace the function body with the compile-time evaluated result is to make the range a function parameter, with a forced instantiation via an alias: [here](https://run.dlang.io/?compiler=ldc&source=import%20std.utf;%0Avoid%20main()%0A%7B%0A%20%20%20%20auto%20m%20%3D%20%22test%22.byCodeUnit;%0A%20%20%20%20int%20res;%0A%09foreach%20(c;%20m)%0A%09%7B%0A%20%20%20%20%20%20%20%20res%20%2B%3D%20c;%0A%09%7D%0A%7D&args=-O).

The function body uses a whole bunch of fancy XMM vectorized operation setup, then a very tight loop summing the values:
````
.cfi_startproc
	testq	%rdi, %rdi
	je	.LBB14_1
	cmpq	$8, %rdi
	jae	.LBB14_4
	xorl	%eax, %eax
	jmp	.LBB14_12
.LBB14_1:
	xorl	%eax, %eax
	retq
.LBB14_4:
	movq	%rdi, %r8
	andq	$-8, %r8
	leaq	-8(%r8), %rdx
	shrq	$3, %rdx
	leal	1(%rdx), %eax
	andl	$1, %eax
	testq	%rdx, %rdx
	je	.LBB14_5
	leaq	-1(%rax), %rcx
	subq	%rdx, %rcx
	pxor	%xmm2, %xmm2
	xorl	%edx, %edx
	pxor	%xmm0, %xmm0
	pxor	%xmm1, %xmm1
	.p2align	4, 0x90
.LBB14_7:
	movd	(%rsi,%rdx), %xmm3
	movd	4(%rsi,%rdx), %xmm4
	punpcklbw	%xmm2, %xmm3
	punpcklwd	%xmm2, %xmm3
	paddd	%xmm0, %xmm3
	punpcklbw	%xmm2, %xmm4
	punpcklwd	%xmm2, %xmm4
	paddd	%xmm1, %xmm4
	movd	8(%rsi,%rdx), %xmm0
	movd	12(%rsi,%rdx), %xmm1
	punpcklbw	%xmm2, %xmm0
	punpcklwd	%xmm2, %xmm0
	paddd	%xmm3, %xmm0
	punpcklbw	%xmm2, %xmm1
	punpcklwd	%xmm2, %xmm1
	paddd	%xmm4, %xmm1
	addq	$16, %rdx
	addq	$2, %rcx
	jne	.LBB14_7
	testq	%rax, %rax
	je	.LBB14_10
.LBB14_9:
	movd	4(%rsi,%rdx), %xmm2
	pxor	%xmm3, %xmm3
	punpcklbw	%xmm3, %xmm2
	punpcklwd	%xmm3, %xmm2
	paddd	%xmm2, %xmm1
	movd	(%rsi,%rdx), %xmm2
	punpcklbw	%xmm3, %xmm2
	punpcklwd	%xmm3, %xmm2
	paddd	%xmm2, %xmm0
.LBB14_10:
	paddd	%xmm1, %xmm0
	pshufd	$78, %xmm0, %xmm1
	paddd	%xmm0, %xmm1
	pshufd	$229, %xmm1, %xmm0
	paddd	%xmm1, %xmm0
	movd	%xmm0, %eax
	cmpq	%r8, %rdi
	je	.LBB14_13
	subq	%r8, %rdi
	addq	%r8, %rsi
	.p2align	4, 0x90
.LBB14_12:
	movzbl	(%rsi), %ecx
	addl	%ecx, %eax
	incq	%rsi
	decq	%rdi
	jne	.LBB14_12
.LBB14_13:
	retq
.LBB14_5:
	pxor	%xmm0, %xmm0
	xorl	%edx, %edx
	pxor	%xmm1, %xmm1
	testq	%rax, %rax
	jne	.LBB14_9
	jmp	.LBB14_10
````

This was with a foreach loop.  With a manual for loop, I get:
````
.cfi_startproc
	testq	%rdi, %rdi
	je	.LBB14_1
	cmpq	$8, %rdi
	jae	.LBB14_4
	xorl	%eax, %eax
	jmp	.LBB14_12
.LBB14_1:
	xorl	%eax, %eax
	retq
.LBB14_4:
	movq	%rdi, %r8
	andq	$-8, %r8
	leaq	-8(%r8), %rdx
	shrq	$3, %rdx
	leal	1(%rdx), %eax
	andl	$1, %eax
	testq	%rdx, %rdx
	je	.LBB14_5
	leaq	-1(%rax), %rcx
	subq	%rdx, %rcx
	pxor	%xmm2, %xmm2
	xorl	%edx, %edx
	pxor	%xmm0, %xmm0
	pxor	%xmm1, %xmm1
	.p2align	4, 0x90
.LBB14_7:
	movd	(%rsi,%rdx), %xmm3
	movd	4(%rsi,%rdx), %xmm4
	punpcklbw	%xmm2, %xmm3
	punpcklwd	%xmm2, %xmm3
	paddd	%xmm0, %xmm3
	punpcklbw	%xmm2, %xmm4
	punpcklwd	%xmm2, %xmm4
	paddd	%xmm1, %xmm4
	movd	8(%rsi,%rdx), %xmm0
	movd	12(%rsi,%rdx), %xmm1
	punpcklbw	%xmm2, %xmm0
	punpcklwd	%xmm2, %xmm0
	paddd	%xmm3, %xmm0
	punpcklbw	%xmm2, %xmm1
	punpcklwd	%xmm2, %xmm1
	paddd	%xmm4, %xmm1
	addq	$16, %rdx
	addq	$2, %rcx
	jne	.LBB14_7
	testq	%rax, %rax
	je	.LBB14_10
.LBB14_9:
	movd	4(%rsi,%rdx), %xmm2
	pxor	%xmm3, %xmm3
	punpcklbw	%xmm3, %xmm2
	punpcklwd	%xmm3, %xmm2
	paddd	%xmm2, %xmm1
	movd	(%rsi,%rdx), %xmm2
	punpcklbw	%xmm3, %xmm2
	punpcklwd	%xmm3, %xmm2
	paddd	%xmm2, %xmm0
.LBB14_10:
	paddd	%xmm1, %xmm0
	pshufd	$78, %xmm0, %xmm1
	paddd	%xmm0, %xmm1
	pshufd	$229, %xmm1, %xmm0
	paddd	%xmm1, %xmm0
	movd	%xmm0, %eax
	cmpq	%r8, %rdi
	je	.LBB14_13
	subq	%r8, %rdi
	addq	%r8, %rsi
	.p2align	4, 0x90
.LBB14_12:
	movzbl	(%rsi), %ecx
	addl	%ecx, %eax
	incq	%rsi
	decq	%rdi
	jne	.LBB14_12
.LBB14_13:
	retq
.LBB14_5:
	pxor	%xmm0, %xmm0
	xorl	%edx, %edx
	pxor	%xmm1, %xmm1
	testq	%rax, %rax
	jne	.LBB14_9
	jmp	.LBB14_10
.Lfunc_end14:
````

It's too long to go through manually, but a diff of the two asm blocks showed that they are instruction-for-instruction identical. IOW, changing from `foreach` to `for` did not make any difference to ldc.